### PR TITLE
Integrate new ScreenBuilder version

### PR DESCRIPTION
**Pull request by Fatme**
This PR fixes 2 problems with ScreenBuilder integration:
1. Currently we are creating files in the current dir - I fixed this behavior and we create directory with appName.
2. Long paths on windows - I moved ScreenBuilder cache directory to <%localappdata%>/Telerik

Should be merged after this https://github.com/telerik/mobile-cli-lib/pull/257

### CODE REVIEW

File: lib/services/app-scaffolding-extensions-service.ts [hunk 1]
```diff
@@ -2,17 +2,19 @@
...
-				$progressIndicator: IProgressIndicator,
-				private $childProcess: IChildProcess,
-		private $dependencyConfigService: IDependencyConfigService) {
-		super($fs, $httpClient, $logger, $progressIndicator);
+		$httpClient: Server.IHttpClient,
+		$logger: ILogger,
+		$progressIndicator: IProgressIndicator,
+		public $childProcess: IChildProcess,
+		public $dependencyConfigService: IDependencyConfigService) {
+		super(options.defaultProfileDir === options.profileDir ? path.join(process.env.LocalAppData, "Telerik") : options.profileDir, $fs, $httpClient, $logger, $progressIndicator);
```
Comment (rosen-vladimirov) [inline lib/services/app-scaffolding-extensions-service.ts]: Can you add a comment why such change is required?

File: lib/helpers.ts [hunk 2]
```diff
@@ -217,3 +219,8 @@ export function fill(value: string, times: number): string[]{
 
 	return repeatedValues;
 }
+
+export function defaultProfileDir(): string {
+	var cacheRootDirectoryPath = hostInfo.isWindows() ? process.env.LocalAppData : path.join(osenv.home(), ".local/share");
+	return path.join(cacheRootDirectoryPath, "Telerik", "BlackDragon", ".appbuilder-cli");
```
Comment (teobugslayer) [inline lib/helpers.ts]: Can we rename .appbuilder to appbuilder? It is in an already hidden dir, so there is no need for it to be hidden as well.

File: lib/services/app-scaffolding-extensions-service.ts [hunk 3]
```diff
@@ -22,9 +24,16 @@ export class AppScaffoldingExtensionsService extends dependencyExtensionsService
...
 			var appScaffoldingConfig = this.$dependencyConfigService.getAppScaffoldingConfig().wait();
-			var afterPrepareAction = () => this.$childProcess.exec("npm install", {cwd: this.appScaffoldingPath });
+			var afterPrepareAction = () => this.afterPrepareAction(this.appScaffoldingPath);
 			this.prepareDependencyExtension(AppScaffoldingExtensionsService.APP_SCAFFOLDING_NAME, appScaffoldingConfig, afterPrepareAction).wait();
 		}).future<void>()();
 	}
+
+	public afterPrepareAction(cwdPath: string):IFuture<void> {
+		return (() => {
+			this.$childProcess.exec("npm install", {cwd: cwdPath }).wait();
```
Comment (teobugslayer) [inline lib/services/app-scaffolding-extensions-service.ts]: Can we use --production here?

File: lib/helpers.ts (hunk 2 above)
Comment (Fatme) [inline lib/helpers.ts]: Actually we can. But this means that we need to move cache directory again. If this is ok, I'll do it.

Comment (rosen-vladimirov) [inline lib/helpers.ts]: I vote to keep it dotted

File: lib/services/app-scaffolding-extensions-service.ts [hunk 4]
```diff
@@ -22,9 +24,22 @@ export class AppScaffoldingExtensionsService extends dependencyExtensionsService
...
 		}).future<void>()();
 	}
+
+	public afterPrepareAction(cwdPath: string, directoryPath?: string):IFuture<void> {
+		return (() => {
+			var command = "npm install ";
+			if(directoryPath) {
+				command += directoryPath;
+			}
+			command += command + " --production";
```
Comment (rosen-vladimirov) [inline lib/services/app-scaffolding-extensions-service.ts]: this way command will be duplicated 
either replace `+=` with `=` or use `command += " --production"`

File: lib/services/screen-builder-service.ts [hunk 5]
```diff
@@ -60,35 +59,51 @@ export class ScreenBuilderService implements IScreenBuilderService {
...
 	}
 
-	private promptGenerate(generatorName: string, type?: string): IFuture<void> {
+	private promptGenerate(generatorName: string, screenBuilderOptions?: IScreenBuilderOptions): IFuture<void> {
+		var scaffolderData = this.createScaffolder(generatorName, screenBuilderOptions);
+		var scaffolder = scaffolderData.scaffolder;
+		var type = screenBuilderOptions.type || ScreenBuilderService.DEFAULT_SCREENBUILDER_TYPE;
+
+		if(type === ScreenBuilderService.DEFAULT_SCREENBUILDER_TYPE) {
+			scaffolder.promptGenerate(type, screenBuilderOptions.answers, scaffolderData.callback);
```
Comment (rosen-vladimirov) [inline lib/services/screen-builder-service.ts]: shouldn't you `.wait()` for this promptGenerate?

Comment (Fatme) [inline lib/services/screen-builder-service.ts]: No, `promptGenerate` comes from Scaffolder API and it doesn't return future.

Comment (rosen-vladimirov) [inline lib/services/screen-builder-service.ts]: aah, correct, I didn't notice it is not our promptGenerate method, sorry :smile:

File: lib/services/app-scaffolding-extensions-service.ts [hunk 6]
```diff
@@ -22,9 +25,25 @@ export class AppScaffoldingExtensionsService extends dependencyExtensionsService
...
+					this.npmDedupe().wait();
+				}).future<void>()();
+			};
 			this.prepareDependencyExtension(AppScaffoldingExtensionsService.APP_SCAFFOLDING_NAME, appScaffoldingConfig, afterPrepareAction).wait();
 		}).future<void>()();
 	}
+
+	protected npmInstall(packageToInstall?: string): IFuture<void> {
+		packageToInstall = packageToInstall || "";
+		var command = util.format("npm install %s %s", packageToInstall, "--production");
```
Comment (teobugslayer) [inline lib/services/app-scaffolding-extensions-service.ts]: The second %s is unneeded:

`npm install ${packageToInstall} --production` ftw!

File: lib/services/app-scaffolding-extensions-service.ts [hunk 7]
```diff
@@ -22,9 +25,25 @@ export class AppScaffoldingExtensionsService extends dependencyExtensionsService
...
+					this.npmDedupe().wait();
+				}).future<void>()();
+			};
 			this.prepareDependencyExtension(AppScaffoldingExtensionsService.APP_SCAFFOLDING_NAME, appScaffoldingConfig, afterPrepareAction).wait();
 		}).future<void>()();
 	}
+
+	protected npmInstall(packageToInstall?: string): IFuture<void> {
+		packageToInstall = packageToInstall || "";
+		var command = util.format("npm install %s ", packageToInstall, "--production");
```
Comment (rosen-vladimirov) [inline lib/services/app-scaffolding-extensions-service.ts]: --production should be moved inside first string:
```TypeScript
var command = util.format("npm install %s --production", packageToInstall);
```

File: lib/services/generator-extensions-service.ts [hunk 8]
```diff
@@ -1,34 +1,44 @@
...
 			var generatorConfig = this.$dependencyConfigService.getGeneratorConfig(generatorName).wait();
-			generatorConfig.pathToSave = this.getGeneratorCachePath(generatorName);
+			var generatorCachePath = this.getGeneratorCachePath(generatorName);
+			generatorConfig.pathToSave = generatorCachePath;
 			this.$fs.ensureDirectoryExists(generatorConfig.pathToSave).wait();
 
-			var afterPrepareAction = () => this.$childProcess.exec("npm install", {cwd: this.getGeneratorCachePath(generatorName) });
+			var afterPrepareAction = () => {
+				return (() => {
+					var dependencies = _.keys(this.$fs.readJson(path.join(generatorCachePath, "package.json")).wait().dependencies);
```
Comment (rosen-vladimirov) [inline lib/services/generator-extensions-service.ts]: You don't need `_.keys` here as on the next line you are using dependencies object, not its keys.

File: lib/helpers.ts (hunk 2 above)
Comment (teobugslayer) [inline lib/helpers.ts]: We discussed this in person and I agree to leave it dotted

File: lib/services/app-scaffolding-extensions-service.ts (hunk 7 above)
Comment (teobugslayer) [inline lib/services/app-scaffolding-extensions-service.ts]: you may remove util.format entirely - just use templated strings (with backquotes).

File: lib/services/extensions-service-base.ts [hunk 9]
```diff
@@ -10,18 +10,15 @@ temp.track();
 export class ExtensionsServiceBase {
 	private extensionVersions: IStringDictionary = {};
 
-	constructor(public $fs: IFileSystem,
-				public $httpClient: Server.IHttpClient,
-				public $logger: ILogger) {
+	constructor(public cacheDir: string,
+		public $fs: IFileSystem,
```
Comment (teobugslayer) [inline lib/services/extensions-service-base.ts]: Why are these public?

File: lib/services/generator-extensions-service.ts (hunk 8 above)
Comment (teobugslayer) [inline lib/services/generator-extensions-service.ts]: If I read the code correctly the "dependencies" are the keys of the "dependencies" - i.e. they are separate entities.

File: lib/services/screen-builder-service.ts [hunk 10]
```diff
@@ -60,35 +59,51 @@ export class ScreenBuilderService implements IScreenBuilderService {
...
 			},
 			logger: this.$logger.trace.bind(this.$logger)
 		};
-		var s = new Scaffolder(connector);
 
+		var scaffolder = new Scaffolder(connector);
 		var future = new Future<void>();
-
-		s.promptGenerate(type, (err:any) => {
+		var callback =  (err:any, data: any) => {
```
Comment (teobugslayer) [inline lib/services/screen-builder-service.ts]: Can we declare err as Error?

File: lib/services/app-scaffolding-extensions-service.ts [hunk 11]
```diff
@@ -2,17 +2,20 @@
...
-				$httpClient: Server.IHttpClient,
-				$logger: ILogger,
-				$progressIndicator: IProgressIndicator,
-				private $childProcess: IChildProcess,
-		private $dependencyConfigService: IDependencyConfigService) {
-		super($fs, $httpClient, $logger, $progressIndicator);
+		$httpClient: Server.IHttpClient,
+		$logger: ILogger,
+		$progressIndicator: IProgressIndicator,
+		public $childProcess: IChildProcess,
```
Comment (teobugslayer) [inline lib/services/app-scaffolding-extensions-service.ts]: Can we make these private or protected?

**Comment by teobugslayer** [comment]
:+1:

File: lib/services/app-scaffolding-extensions-service.ts [hunk 12]
```diff
@@ -2,17 +2,20 @@
...
-				$progressIndicator: IProgressIndicator,
-				private $childProcess: IChildProcess,
-		private $dependencyConfigService: IDependencyConfigService) {
-		super($fs, $httpClient, $logger, $progressIndicator);
+		$httpClient: Server.IHttpClient,
+		$logger: ILogger,
+		$progressIndicator: IProgressIndicator,
+		protected $childProcess: IChildProcess,
+		protected $dependencyConfigService: IDependencyConfigService) {
+		super(options.defaultProfileDir === options.profileDir ? path.join(process.env.LocalAppData, "Telerik", "sb") : options.profileDir, $fs, $httpClient, $logger, $progressIndicator); // We should pass here the correct profileDir
```
Comment (rosen-vladimirov) [inline lib/services/app-scaffolding-extensions-service.ts]: you can use `process.env.LocalAppData` only on Windows :(

File: lib/services/app-scaffolding-extensions-service.ts [hunk 13]
```diff
@@ -22,9 +25,26 @@ export class AppScaffoldingExtensionsService extends dependencyExtensionsService
 	public prepareAppScaffolding(): IFuture<void> {
 		return (() => {
 			var appScaffoldingConfig = this.$dependencyConfigService.getAppScaffoldingConfig().wait();
-			var afterPrepareAction = () => this.$childProcess.exec("npm install", {cwd: this.appScaffoldingPath });
+			appScaffoldingConfig.pathToSave = options.defaultProfileDir === options.profileDir ? path.join(process.env.LocalAppData, "Telerik", "sb") : options.profileDir;
```
Comment (rosen-vladimirov) [inline lib/services/app-scaffolding-extensions-service.ts]: you can use process.env.LocalAppData only on Windows :(

**Comment by teobugslayer** [comment]
:+1:

**Comment by rosen-vladimirov** [comment]
:+1:  great work :smile_cat:

**Pull request merged**